activestorage/app/controllers/active_storage/direct_uploads_controller.rb

```diff
@@ -5,13 +5,17 @@
 # the blob that was created up front.
 class ActiveStorage::DirectUploadsController < ActiveStorage::BaseController
   def create
-    blob = ActiveStorage::Blob.create_before_direct_upload!(**blob_args)
+    blob = ActiveStorage::Blob.create_before_direct_upload!(**blob_args.merge(service_name: verified_service_name))
     render json: direct_upload_json(blob)
   end
 
   private
     def blob_args
-      params.require(:blob).permit(:filename, :byte_size, :checksum, :content_type, :metadata).to_h.symbolize_keys
+      params.require(:blob).permit(:filename, :byte_size, :checksum, :content_type, :metadata, :service_name).to_h.symbolize_keys
+    end
+
+    def verified_service_name
```

Comment:
I'm implementing your changes into my own app, and I'm running into issues with this method. This keeps returning an `ActiveSupport::MessageVerifier::InvalidSignature` exception. I feel we might not need this because there's [a validation on the service_name](https://github.com/rails/rails/blob/master/activestorage/app/models/active_storage/blob.rb#L66) on the Blob object.

Reply:
The validation you mentioned ensures that service with this name _exists_, while my validation decodes the [encrypted](https://github.com/rails/rails/pull/38957/files#diff-35ff9932cd62aff4f22fa9fb8fc9ea39005108d86e8854c8d7a9567cc3871bc7R922) service name. We need it to prevent situation, when there is an attachment X (can be directly uploaded) and Y (can be directly uploaded) and user can upload anything he wants to Y if he knows the service name. With the encrypted name he has to find out the encrypted service name first.